updates to booting RHCOS via PXE (#27120)
openshift/openshift-docs: modules/installation-user-infra-machines-pxe.adoc

The file names contain the {product-title} version number.
They resemble the following examples:

** Compressed metal RAW image: `rhcos-<version>-<architecture>-metal.<architecture>.raw.gz`
** `kernel`: `rhcos-<version>-<architecture>-installer-kernel-<architecture>`
** `initramfs`: `rhcos-<version>-<architecture>-installer-initramfs.<architecture>.img`
endif::openshift-origin[]
ifdef::openshift-origin[]
. Obtain the {op-system} images from the
link:https://getfedora.org/en/coreos/download?tab=metal_virtualized&stream=stable[{op-system} Downloads] page
endif::openshift-origin[]

> . If you are using traditional PXE, you should upload the `kernel` and `initramfs` files to your TFTP server.  If you are using iPXE, you should upload the `kernel` and `initramfs` files to your HTTP server.  In both cases, you should upload the RAW image to your HTTP server.

Reviewer: ```suggestion
. Upload the RAW image to your HTTP server.

. Upload the additional files that are required for your booting method:
* For traditional PXE, upload the `kernel` and `initramfs` files to your TFTP server.
* For iPXE, upload the `kernel` and `initramfs` files to your HTTP server. 
```

initrd http://<HTTP_server>/rhcos-<version>-<architecture>-installer-initramfs.<architecture>.img <3>
boot
----
<1> Specify locations of the {op-system} files that you uploaded to your
HTTP server. The `kernel` parameter value is the location of the `kernel` file,
the `initrd` parameter value references the name of the `initramfs` file that is
supplied on the `initrd` line below, the `coreos.inst.image_url` parameter value
is the location of the compressed metal RAW image, and the `coreos.inst.ignition_url`
parameter value is the location of the bootstrap Ignition config file.
<2> If you use multiple NICs, specify a single interface in the `ip` option.
For example, to use DHCP on a NIC that is named `eno1`, set `ip=eno1:dhcp`.
<3> Specify the location of the `initramfs` file that you uploaded to your HTTP
server.
endif::only-pxe[]

> . If you use UEFI, you will need to provide the necessary EFI binaries and `grub.cfg`

Reviewer: I think that the new content needs to go in a single substep, like this:

````
. If you use UEFI, perform the following actions:
.. Provide the EFI binaries and `grub.cfg` file that are required for booting the system. You also need the `shim.efi` binary and the `grubx64.efi` binary. Use one of the following methods to obtain these files:

** Retrieve these files from the `shim` and `grub2-efi` packages included with Red Hat Enterprise Linux 8.  See the link:https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/8/html/performing_an_advanced_rhel_installation/preparing-for-a-network-install_installing-rhel-as-an-experienced-user#configuring-a-tftp-server-for-uefi-based-clients_preparing-for-a-network-install[RHEL8 instructions] for extracting the EFI binaries from RPMs.

** Extract the necessary EFI binaries by mounting the {op-system} ISO on
your host, copying the `images/efiboot.img` file, then mounting `efiboot.img` on your host. From the `efiboot.img` mount point, you then copy the `EFI/redhat/shimx64.efi` and
`EFI/redhat/grubx64.efi` files to your TFTP server.

.. Copy the `EFI/redhat/grub.cfg` file that is included in the {op-system} ISO to
the necessary location on your TFTP server. 

.. Edit the `grub.cfg` file to include the following arguments:
+
----
menuentry 'Install Red Hat Enterprise Linux CoreOS' --class fedora --class gnu-linux --class gnu --class os {
	linux rhcos-<version>-<architecture>-installer-kernel-<architecture> nomodeset rd.neednet=1 coreos.inst=yes coreos.inst.install_dev=sda coreos.inst.image_url=http://<HTTP_server>/rhcos-<version>-<architecture>-metal.<architecture>.raw.gz coreos.inst.ignition_url=http://<HTTP_server>/bootstrap.ign <1>
	initrd rhcos-<version>-<architecture>-installer-initramfs.<architecture>.img <2>
}
----
<1> The first argument to the `linux` line item is the location of the `kernel`
file that you uploaded to your TFTP server. For the `coreos.inst.image_url` parameter
value, specify the location of the compressed metal RAW image that you uploaded
to your HTTP server. For the `coreos.inst.ignition_url` paramter, specify the location of the
bootstrap Ignition config file that you uploaded to your HTTP server.
<2> Specify the location of the `initramfs` file that you uploaded to your TFTP
server.
````

required for booting the system.  You will need the `shim.efi` binary and the `grubx64.efi`
binary.  These can be retrieved from the `shim` and `grub2-efi` packages included with
Red Hat Enterprise Linux 8.  See the link:https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/8/html/performing_an_advanced_rhel_installation/preparing-for-a-network-install_installing-rhel-as-an-experienced-user#configuring-a-tftp-server-for-uefi-based-clients_preparing-for-a-network-install[RHEL8 instructions]
for how you can extract the EFI binaries from RPMs.

> . Alternately, you can extract the necessay EFI binaries by mounting the {op-system} ISO on

Reviewer 2: ```suggestion
. Alternately, you can extract the necessary EFI binaries by mounting the {op-system} ISO on
```

HTTP or TFTP server. The `initrd` parameter value is the location of the `initramfs`
file on your TFTP server. The `coreos.inst.image_url` parameter value is the
location of the compressed metal RAW image on your HTTP server, and the
`coreos.inst.ignition_url` parameter value is the location of the bootstrap
Ignition config file on your HTTP server.

ifndef::only-pxe[]
** For iPXE:
+
----
kernel  http://<HTTP_server>/rhcos-<version>-<architecture>-installer-kernel-<architecture> ip=dhcp rd.neednet=1 initrd=rhcos-<version>-<architecture>-installer-initramfs.<architecture>.img console=tty0 console=ttyS0 coreos.inst=yes coreos.inst.install_dev=sda coreos.inst.image_url=http://<HTTP_server>/rhcos-<version>-<architecture>-metal.<architecture>.raw.gz coreos.inst.ignition_url=http://<HTTP_server>/bootstrap.ign <1> <2>
initrd http://<HTTP_server>/rhcos-<version>-<architecture>-installer-initramfs.<architecture>.img <3>
boot
----
<1> Specify locations of the {op-system} files that you uploaded to your
HTTP server. The `kernel` parameter value is the location of the `kernel` file,
the `initrd` parameter value references the name of the `initramfs` file that is

> supplied on the `initrd` line below, the `coreos.inst.image_url` parameter value

Reviewer: ```suggestion
supplied on the following `initrd` line, the `coreos.inst.image_url` parameter value
```

is the location of the compressed metal RAW image, and the `coreos.inst.ignition_url`
parameter value is the location of the bootstrap Ignition config file.
<2> If you use multiple NICs, specify a single interface in the `ip` option.
For example, to use DHCP on a NIC that is named `eno1`, set `ip=eno1:dhcp`.
<3> Specify the location of the `initramfs` file that you uploaded to your HTTP
server.
endif::only-pxe[]

. If you use UEFI, perform the following actions:
.. Provide the EFI binaries and `grub.cfg` file that are required for booting the system. You need the `shim.efi` binary and the `grubx64.efi` binary.

** Extract the necessary EFI binaries by mounting the {op-system} ISO on
your host and then mounting the `images/efiboot.img` file to your host.
From the `efiboot.img` mount point, you then copy the `EFI/redhat/shimx64.efi` and
`EFI/redhat/grubx64.efi` files to your TFTP server.
+
----
# mkdir -p /mnt/{iso,efiboot}
# mount -o loop rhcos-installer.x86_64.iso /mnt/iso

> # mount -o loop,ro /mnt/iso/images/efiboot.imgefiboot.img /mnt/efiboot

Reviewer 3: `efiboot.imgefiboot.img`

Author: Good eyes!

> # mount -o loop  -t vfat efiboot.img /mnt/efiboot

Reviewer 3: We could simplify this by skipping the `cp` and doing `mount -o loop,ro` directly from `/mnt/iso/images/efiboot.img`.  The summary description above would need to be updated also.

Author: Good suggestion!

initrd http://<HTTP_server>/rhcos-<version>-<architecture>-installer-initramfs.<architecture>.img <3>
boot
----
<1> Specify locations of the {op-system} files that you uploaded to your
HTTP server. The `kernel` parameter value is the location of the `kernel` file,
the `initrd` parameter value references the name of the `initramfs` file that is
supplied on the `initrd` line below, the `coreos.inst.image_url` parameter value
is the location of the compressed metal RAW image, and the `coreos.inst.ignition_url`
parameter value is the location of the bootstrap Ignition config file.
<2> If you use multiple NICs, specify a single interface in the `ip` option.
For example, to use DHCP on a NIC that is named `eno1`, set `ip=eno1:dhcp`.
<3> Specify the location of the `initramfs` file that you uploaded to your HTTP
server.
endif::only-pxe[]

. If you use UEFI, perform the following actions:

> .. Provide the EFI binaries and `grub.cfg` file that are required for booting the system. You need the `shim.efi` binary and the `grubx64.efi` binary. Use one of the following methods to obtain these files:

Reviewer 3: The existing PXE/iPXE instructions assume the DHCP server is already configured to point to PXELINUX or iPXE, which might be okay because setup instructions are widely available.  Is it similarly safe to assume that the customer knows how to point DHCP to shim?

Author: We could add some language to the prerequisites, etc that says something like "These instructions assume you are familiar with configuring DHCP/PXE infrastructure."

Author: Pushed an update with an edit to the prerequisite section

> ** Retrieve these files from the `shim` and `grub2-efi` packages included with Red Hat Enterprise Linux 8.  See the link:https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/8/html/performing_an_advanced_rhel_installation/preparing-for-a-network-install_installing-rhel-as-an-experienced-user#configuring-a-tftp-server-for-uefi-based-clients_preparing-for-a-network-install[RHEL8 instructions] for extracting the EFI binaries from RPMs. Then copy the `shim.efi` and `grubx64.efi` binaries to your TFTP server.

Reviewer 3: Since we're asking the user to copy the `grub.cfg` from the ISO anyway, maybe we should drop this case?

Author: @kalexand-rh Do you think there is value in providing additional ways to accomplish the end goal or keep it simple?

Reviewer: @miabbott, unless there's a clear use case for why you'd need a more complicated method, I think that it's best to present the simplest method.

Author: OK, let me remove it and push a final (:crossed_fingers:) update

** Extract the necessary EFI binaries by mounting the {op-system} ISO on
your host, copying the `images/efiboot.img` file to your host, then mounting `efiboot.img` on your host. From the `efiboot.img` mount point, you then copy the `EFI/redhat/shimx64.efi` and

> `EFI/redhat/grubx64.efi` files to your TFTP server.

Reviewer 3: Optional: provide sample commands for extracting the files.

Author: Pushed an update with examples

.. Copy the `EFI/redhat/grub.cfg` file that is included in the {op-system} ISO to your TFTP server.

.. Edit the `grub.cfg` file to include the following arguments:
+
----
menuentry 'Install Red Hat Enterprise Linux CoreOS' --class fedora --class gnu-linux --class gnu --class os {

> 	linux rhcos-<version>-<architecture>-installer-kernel-<architecture> nomodeset rd.neednet=1 coreos.inst=yes coreos.inst.install_dev=sda coreos.inst.image_url=http://<HTTP_server>/rhcos-<version>-<architecture>-metal.<architecture>.raw.gz coreos.inst.ignition_url=http://<HTTP_server>/bootstrap.ign <1>

Reviewer 4: if they use nvme SSD, the install_dev still be sda?

Author: The disk can change based on the disk type or number of disks.  This is just an example.
